### PR TITLE
fix incorrect coverage values in the compare test coverage workflow

### LINKED ISSUE #826: Incorrect coverage values in the `Compare Test Coverage` workflow
[Context](https://github.com/ubiquity/ubiquity-dollar/pull/825#issuecomment-1777205152)

Check [this](https://github.com/ubiquity/ubiquity-dollar/actions/runs/6627359435/job/18002118554#logs) CI run which calculates the following coverage values and fails afterwards because the coverage has been decreased:
```
Development branch coverage: 69.1
PR branch coverage: 10.1
```

The PR branch coverage couldn't fail so drastically. We should find the root cause of the bug and fix it.

### PR DESCRIPTION
**Pull request by gitcoindev**
- fetch development and pull request branches to separate folders
- set working-directory during get coverage steps
- forge build before forge coverage to be able to access JSON ABI facet contracts

Resolves: #826

### CODE REVIEW

**Comment by gitcoindev** [comment]
I tested multiple times on my fork and just added deleting development branch folder after calculating coverage to really make sure that coveralls does not read anything from it. Should not happen, but adding this step increases safety.

New QA: https://coveralls.io/jobs/130726799

I propose to review and close this pull request 'as is' as it fixes the bug and open a new edge case issue only when it appears.

It is additionally possible to set pull request alerts on coveralls side if needed, e.g. if 70% or any arbitrary coverage is considered 'green' and anything below should be marked as 'red' in the badge @pavlovcik @rndquu @molecula451 .

image

**Comment by rndquu** [comment]
@gitcoindev Could you provide a QA CI run when the "Compare Test Coverage" fails on decreased test coverage?

**Comment by gitcoindev** [comment]
@rndquu sure.

Step 1 - I synchronized my ubiquity-dollar fork development branch 1:1 to match upstream.
Step 2 - I opened a pull request https://github.com/gitcoindev/ubiquity-dollar/pull/7 which is a mirror of this pull request, but directed towards my forked repository.
Step 3 - I did QA of my mirror pull request here https://github.com/gitcoindev/ubiquity-dollar/actions/runs/6651769749 . I received a comment that proves that coverage percentage was fixed https://github.com/gitcoindev/ubiquity-dollar/pull/7#issuecomment-1780707064
Step 4 - I merged my mirror pull request to simulate merging this one.
Step 5 - I removed  test cases to reduce coverage and opened an another pull request to my fork https://github.com/gitcoindev/ubiquity-dollar/pull/8

Result: coverage decreased and job is failing. QA: https://github.com/gitcoindev/ubiquity-dollar/actions/runs/6652771962/job/18077404702

**Comment by gitcoindev** [comment]
image

**Review by rndquu — APPROVED** [review]

File: .github/workflows/coverage-check.yml [hunk 1]
```diff
@@ -59,28 +59,25 @@ jobs:
...
         id: coverage-pr
+        working-directory: packages/contracts
         run: |
-          cd ./packages/contracts
-
           # generates lcov.info
-          forge coverage --report lcov
+          forge build && forge coverage --report lcov
 
           # Foundry uses relative paths but Hardhat uses absolute paths.
```
Comment (molecula451) [inline .github/workflows/coverage-check.yml]: @gitcoindev can you update this to have it good on this same PR? we do not use hardhat on this repo, appreciate it

File: .github/workflows/coverage-check.yml [hunk 2]
```diff
@@ -30,11 +31,10 @@ jobs:
...
         id: coverage-development
+        working-directory: development/packages/contracts
         run: |
-          cd ./packages/contracts
-
           # generates lcov.info
-          forge coverage --report lcov
+          forge build && forge coverage --report lcov
 
           # Foundry uses relative paths but Hardhat uses absolute paths.
```
Comment (molecula451) [inline .github/workflows/coverage-check.yml]: here too

File: .github/workflows/coverage-check.yml (hunk 1 above)
Comment (gitcoindev) [reply .github/workflows/coverage-check.yml]: Done, see commit 5b754a4d03f65f88b8ffde8b95e978a8358ea381

**Comment by gitcoindev** [comment]
@molecula451 I applied required changes, and would be grateful if you had a look again at the pr.

**Comment by molecula451** [comment]
the improvement looks good, but, lets test it a bit more, I have re-ran rndqnuu PR's action with the old workflow and it seems to be greening now, any thoughts on why? https://github.com/ubiquity/ubiquity-dollar/actions/runs/6627359435/workflow you can check #attempt 2

**Comment by gitcoindev** [comment]
> the improvement looks good, but, lets test it a bit more, I have re-ran rndqnuu PR's action with the old workflow and it seems to be greening now, any thoughts on why? https://github.com/ubiquity/ubiquity-dollar/actions/runs/6627359435/workflow you can check #attempt 2

@molecula451  yes, it is greenish but my thoughts are: it is broken. 

Development branch coverage: 10.1
PR branch coverage: 10.1

Should be > 60%. The job is green and passing only because it does not show coverage degradation, but the calculated coverage value itself is wrong. Please have a look at the compared coverage values.

**Comment by gitcoindev** [comment]
Current correct values:

Development branch coverage: 69.1
PR branch coverage: 69.1

**Review by molecula451 — APPROVED** [review]
our branch is holding 69.1% coverage on coveralls, this PR fix missing values that were part of last PR and needed to be update, our current branch only solves 10% coverage which is not enough to gain us a broader perspective, this PR solves, updates, and improves the coverage to get back to our 69.1% standard value

**Comment by gitcoindev** [comment]
Thank you for approvals and merging!

**Pull request merged**